### PR TITLE
RmsProp stochastic optimization in Neural Nets

### PR DESCRIPTION
**Pull request by arasuarun**
From issue #2967 

Implemented RmsProp training algorithm from optimization/RmsPropUpdater.h to Neural Networks, in NeuralNetwork.cpp.
Default values of the hyper parameters are the standard default values: epsilon=1e-6 and decay_factor=0.9.

I also wrote a simple unit test to make sure rmsprop works, based off the regression unit test.

### CODE REVIEW

**Comment by arasuarun** [comment]
In the gradient_descent unit test, I set the values of epsilon and decay_factor as 1.0 and 0.99 respective. This ensures that rmsprop has minimal effect in learning and doesn't interfere.

Ideally, decay_factor should be 1.0, but that's not permitted. I will later send a PR proposing to change its range from [0,1) to [0,1], if you guys will allow that.

File: src/shogun/neuralnets/NeuralNetwork.cpp [hunk 1]
```diff
@@ -265,6 +266,10 @@ bool CNeuralNetwork::train_gradient_descent(SGMatrix<float64_t> inputs,
 		"Gradient descent learning rate (%f) must be > 0\n", gd_learning_rate);
 	REQUIRE(gd_momentum>=0,
 		"Gradient descent momentum (%f) must be > 0\n", gd_momentum);
+  REQUIRE(rms_epsilon>=0,
```
Comment (lisitsyn) [inline src/shogun/neuralnets/NeuralNetwork.cpp]: Here and through the changes, new code's indentation doesn't match original one.

Comment (arasuarun) [inline src/shogun/neuralnets/NeuralNetwork.cpp]: I think all of those should be fixed now. My tab length in vim was too small and I didn't notice the tab/space issue.

File: src/shogun/neuralnets/NeuralNetwork.h [hunk 2]
```diff
@@ -442,6 +442,18 @@ friend class CDeepBeliefNetwork;
 	 * default value is -1
 	 */
 	float64_t gd_error_damping_coeff;
+
+	/** Used in RmsProp stochastic optimization
```
Comment (iglesias) [inline src/shogun/neuralnets/NeuralNetwork.h]: I **think** you need to use [this syntax](https://www.stack.nl/~dimitri/doxygen/manual/grouping.html#memgroup) to group the documentation of more than one member together.

File: src/shogun/neuralnets/NeuralNetwork.h [hunk 3]
```diff
@@ -442,6 +442,18 @@ friend class CDeepBeliefNetwork;
...
 	 */
 	float64_t gd_error_damping_coeff;
+
+	/** Used in RmsProp stochastic optimization 
+	* RmsProp scales each gradient down by its running average.
+	* It then updates the parameters acc. to the learning_rate.
+	*     scale=decay_factor*scale+(1-decay_factor)*gradient*gradient
+	*     param=param-learning_rate*gradient/sqrt(scale+epsilon)
+	* @epsilon: default=1e-6
+	* @decay_factor: default=0.9
```
Comment (iglesias) [inline src/shogun/neuralnets/NeuralNetwork.h]: You don't need the @'s here, I believe.

File: src/shogun/neuralnets/NeuralNetwork.h [hunk 4]
```diff
@@ -442,6 +442,18 @@ friend class CDeepBeliefNetwork;
 	 * default value is -1
 	 */
 	float64_t gd_error_damping_coeff;
+
+	/** Used in RmsProp stochastic optimization 
+	* RmsProp scales each gradient down by its running average.
+	* It then updates the parameters acc. to the learning_rate.
+	*     scale=decay_factor*scale+(1-decay_factor)*gradient*gradient
```
Comment (iglesias) [inline src/shogun/neuralnets/NeuralNetwork.h]: Can we get these equations with Latex? See for instance the doc of check_gradients in this file.

Comment (arasuarun) [inline src/shogun/neuralnets/NeuralNetwork.h]: Okay. I'll fix all of those right away.

File: src/shogun/neuralnets/NeuralNetwork.cpp [hunk 5]
```diff
@@ -264,7 +265,11 @@ bool CNeuralNetwork::train_gradient_descent(SGMatrix<float64_t> inputs,
 	REQUIRE(gd_learning_rate>0,
 		"Gradient descent learning rate (%f) must be > 0\n", gd_learning_rate);
 	REQUIRE(gd_momentum>=0,
-		"Gradient descent momentum (%f) must be > 0\n", gd_momentum);
+    		"Gradient descent momentum (%f) must be > 0\n", gd_momentum);
```
Comment (iglesias) [inline src/shogun/neuralnets/NeuralNetwork.cpp]: I am not sure why the change in this line. Anyway, maybe we can fix the > to >= as the actual require is doing :-)

Comment (arasuarun) [inline src/shogun/neuralnets/NeuralNetwork.cpp]: I actually wanted to change it. But since it was completely unrelated to what I was actually doing, I decided to save it for another PR.

**Comment by iglesias** [comment]
If would be great if you can squash the indentation commits into the main one.

**Comment by arasuarun** [comment]
I fixed up:
1. proper documentation for grouping members together
2. using Latex for the formulae
3. remove some unnecessary '@'s

I'm sorry for the trouble - it's my first PR and I was careless. 
I'll now get onto squashing the 5 commits into 1. 
Thanks :)

**Comment by iglesias** [comment]
Thanks to you for getting back with the fixes swiftly :-)

File: src/shogun/neuralnets/NeuralNetwork.h [hunk 6]
```diff
@@ -442,6 +442,20 @@ friend class CDeepBeliefNetwork;
 	 * default value is -1
 	 */
 	float64_t gd_error_damping_coeff;
+
+	//@{
```
Comment (lisitsyn) [inline src/shogun/neuralnets/NeuralNetwork.h]: I am not sure about this thing - what is it for?

**Comment by yorkerlin** [comment]
@lisitsyn @iglesias @arasuarun 
We should use an interface for optimizers instead of a specific optimizer (eg, `rmsprop`). 
Please consider the following case:
if users want to use ADAM/ADGRAD optimizer, it seems that we have to modify the code according to this PR. 
@arasuarun  correct me if I am wrong.

File: src/shogun/neuralnets/NeuralNetwork.h (hunk 6 above)
Comment (arasuarun) [inline src/shogun/neuralnets/NeuralNetwork.h]: Grouping similar members together. Both of those are only used for RmsProp.  https://www.stack.nl/%7Edimitri/doxygen/manual/grouping.html#memgroup

**Comment by arasuarun** [comment]
@yorkerlin That's completely right. Now that you mention it, since they all derive from the class DescendUpdaterWithCorrection and have the same members/interface, it makes more sense to have a general implementation.

I raised #2967 some time back (and didn't get any response, so I went ahead with it :P ). I think we can continue the discussion there.

**Comment by karlnapf** [comment]
news?

**Comment by arasuarun** [comment]
Sorry for the late response @karlnapf, I missed your ping among the various the notifications. I now get a taste of how you guys must feel. :) 

Reg. this, considering we have PR #2973, this is no longer required. I'll go ahead and close this. Thanks.
